File: plumber/gateway/endpoints/deleteBot.R
##########################################################################
#
#* @param apiKey The API key for authentication
#* @param protocol The protocol to use
#* @param network The network to use
#* @param pool The pool to target
#* @response 200 Returns the result of the bot deletion
#* @response 400 Bad request
#* @response 500 Internal server error
#* @tag managers
#* @delete /managers/deleteBot
#
##########################################################################

deleteBotHandler <- function(apiKey = NULL, protocol = "dhedge", network = NULL, pool = NULL) {
  protocol <- tolower(protocol)
  pool     <- tolower(pool)
  network  <- tolower(network)

  # === Validate required params ===
  required_params <- list(apiKey = apiKey, network = network, pool = pool)
  missing <- names(required_params)[sapply(required_params, function(x) is.null(x) || x == "")]
  if (length(missing) > 0) {
    return(list(
      status = "fail",
      status_code = 400,
      message = paste("Missing required parameters:", paste(missing, collapse = ", "))
    ))
  }

  # === Run basic validation ===
  check <- basic_check(network = network, protocol = protocol, pool = pool, apiKey = apiKey)
  if (!is.null(check$status) && check$status == "fail") {
    return(list(
      status = "fail",
      status_code = 400,
      message = check$message
    ))
  }

  # === Build the request URL ===
  url <- paste0(pep, "deleteBot?",
                "apiKey=", apiKey,
                "&protocol=", protocol,
                "&pool=", pool,
                "&network=", network)

  # === Make the request safely ===
  result <- tryCatch({
    response <- httr::POST(url)
    code <- httr::status_code(response)
    text <- httr::content(response, "text", encoding = "UTF-8")
    parsed <- tryCatch(jsonlite::fromJSON(text), error = function(e) list(message = text))

    msg <- if (!is.null(parsed$message)) parsed$message else text

    if (code >= 200 && code < 300) {
      list(status = "success", status_code = code, message = msg)
    } else {
      list(status = "fail", status_code = code, message = msg)
    }
  }, error = function(e) {
    list(status = "fail", status_code = 500, message = paste("Internal server error:", e$message))
  })

  return(result)
}

pr$handle("DELETE","/deleteBot",deleteBotHandler, comment="This endpoint is used to turn off the trading bot.")

File: infinitetrading_api/express/src/requests/deleteBot.ts
/**
 * requests/deleteBot.ts — Node port of R gateway's deleteBot.R (proxies to
 * plumber-api port 8002's deleteBotHandler -> db.R's deleteBot()).
 *
 * PARITY NOTES:
 *  - Gateway layer: requires apiKey, network, pool (protocol defaults to
 *    "dhedge") -- missing any -> 400 "Missing required parameters: ...".
 *    Then basic_check(network, protocol, pool, apiKey) (UUID apiKey scheme,
 *    same family as getGasBalance/getGasWalletPools) -- fail -> 400 with
 *    check's own message (gateway wraps ALL basic_check failures as generic
 *    400, not the original status_code -- confirmed from R source:
 *    `if (!is.null(check$status) && check$status == "fail") return(list(status="fail",status_code=400,message=check$message))`).
 *  - Inner (port 8002) layer: isValidApiKey(network, protocol, pool, apiKey)
 *    -- a DB check (SELECT 1 FROM gas_wallets WHERE network=? AND protocol=?
 *    AND pool=? AND token=?) -- fail -> 401 "Invalid API key" (note: this is
 *    the SAME apiKey validated twice, first as a UUID format then as a real
 *    linked gas-wallet token -- both must pass).
 *  - deleteBot() itself: DELETE FROM dhedge_sides WHERE network=? AND pool=?
 *    (protocol is accepted but NOT used in the WHERE clause -- confirmed
 *    from R source, deletes across all protocols for that network+pool).
 *  - Response is a straight pass-through in R (gateway wraps the port-8002
 *    response in its own success/fail/message shape from the raw JSON body)
 *    -- replicated by returning success/status_code/message directly from
 *    the single, combined check+delete flow (no real inter-process hop
 *    needed once both R hops collapse into one Express handler).
 */

import { Router, Request, Response } from 'express';
import { dbQuery } from '../db';
import { basicCheck } from '../basicCheck';

const router = Router();

async function isValidApiKeyForPool(network: string, protocol: string, pool: string, apiKey: string): Promise<boolean> {
  try {
    const rows = await dbQuery(
      'SELECT 1 FROM gas_wallets WHERE network = ? AND protocol = ? AND pool = ? AND token = ? LIMIT 1',
      [network.toLowerCase(), protocol, pool.toLowerCase(), apiKey]
    );
    return rows.length > 0;
  } catch (e: any) {
    console.log('Error: isValidApiKeyForPool:', e.message);
    return false;
  }
}

/**
 * @openapi
 * /deleteBot:
 *   delete:
 *     summary: Delete a trading bot's side configuration for a pool (turns off the bot)
 *     tags: [Managers]
 *     parameters:
 *       - in: query
 *         name: apiKey
 *         required: true
 *         schema: { type: string }
 *       - in: query
 *         name: protocol
 *         schema: { type: string, default: dhedge }
 *       - in: query
 *         name: network
 *         required: true
 *         schema: { type: string }
 *       - in: query
 *         name: pool
 *         required: true
 *         schema: { type: string }
 *     responses:
 *       200:
 *         description: Bot deleted.
 */
router.delete('/deleteBot', async (req: Request, res: Response) => {
  const apiKey = String(req.query.apiKey || '');
  const protocolRaw = String(req.query.protocol || 'dhedge').toLowerCase();
  const networkRaw = String(req.query.network || '').toLowerCase();
  const poolRaw = String(req.query.pool || '').toLowerCase();

  const missing: string[] = [];
  if (!apiKey) missing.push('apiKey');
  if (!networkRaw) missing.push('network');
  if (!poolRaw) missing.push('pool');
  if (missing.length > 0) {
    return res.json({ status: ['fail'], status_code: [400], message: [`Missing required parameters: ${missing.join(', ')}`] });
  }

  const check = await basicCheck({ network: networkRaw, protocol: protocolRaw, pool: poolRaw, apiKey });
  if (check.status === 'fail') {
    // Gateway layer wraps ANY basic_check failure as a generic 400 with the
    // check's own message (not the original status_code).
    return res.json({ status: ['fail'], status_code: [400], message: [check.message] });
  }

  const validForPool = await isValidApiKeyForPool(networkRaw, protocolRaw, poolRaw, apiKey);
  if (!validForPool) {
    return res.json({ status: ['fail'], status_code: [401], message: ['Invalid API key'] });
  }

  try {
    await dbQuery('DELETE FROM dhedge_sides WHERE network = ? AND pool = ?', [networkRaw, poolRaw]);
    return res.json({ status: ['success'], status_code: [200], message: [`Bot successfully deleted: ${poolRaw}`] });
  } catch (e: any) {
    console.log(`Error: deleteBot for protocol: ${protocolRaw} network: ${networkRaw} pool: ${poolRaw} error: ${e.message}`);
    return res.json({ status: ['fail'], status_code: [500], message: ['Internal error deleting bot entry'] });
  }
});

export default router;
